common/tasks/sequential_scheduler.go

```diff
@@ -0,0 +1,333 @@
+			return nil
+		},
+	)
+	if err != nil {
+		panic("Error is not expected as the evaluation function returns nil")
+	}
+
+	// if function evaluated, meaning that the task set is
+	// already dispatched
+	if fnEvaluated {
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return
+	}
+
+	// need to dispatch this task set
+	select {
+	case <-s.shutdownChan:
+		task.Cancel()
+	case s.queueChan <- queue:
+		if s.isStopped() {
+			s.drainTasks()
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) TrySubmit(task T) bool {
+	queue := s.queueFactory(task)
+	queue.Add(task)
+
+	_, fnEvaluated, err := s.queues.PutOrDo(
+		queue.ID(),
+		queue,
+		func(key interface{}, value interface{}) error {
+			value.(SequentialTaskQueue[T]).Add(task)
+			return nil
+		},
+	)
+	if err != nil {
+		panic("Error is not expected as the evaluation function returns nil")
+	}
+	if fnEvaluated {
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return true
+	}
+
+	select {
+	case s.queueChan <- queue:
+		if s.isStopped() {
+			s.drainTasks()
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *SequentialScheduler[T]) workerMonitor() {
+	defer s.shutdownWG.Done()
+
+	for {
+		select {
+		case <-s.shutdownChan:
+			s.stopWorkers(len(s.workerShutdownCh))
+			return
+		case <-time.After(backoff.Jitter(defaultMonitorTickerDuration, defaultMonitorTickerJitter)):
+			targetWorkerNum := s.options.WorkerCount()
+			currentWorkerNum := len(s.workerShutdownCh)
+
+			if targetWorkerNum == currentWorkerNum {
+				continue
+			}
+
+			if targetWorkerNum > currentWorkerNum {
+				s.startWorkers(targetWorkerNum - currentWorkerNum)
+			} else {
+				s.stopWorkers(currentWorkerNum - targetWorkerNum)
+			}
+			s.logger.Info("Update worker pool size", tag.Key("worker-pool-size"), tag.Value(targetWorkerNum))
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) startWorkers(
+	count int,
+) {
+	for i := 0; i < count; i++ {
+		shutdownCh := make(chan struct{})
+		s.workerShutdownCh = append(s.workerShutdownCh, shutdownCh)
+
+		s.shutdownWG.Add(1)
+		go s.pollTaskQueue(shutdownCh)
+	}
+}
+
+func (s *SequentialScheduler[T]) stopWorkers(
+	count int,
+) {
+	shutdownChToClose := s.workerShutdownCh[:count]
+	s.workerShutdownCh = s.workerShutdownCh[count:]
+
+	for _, shutdownCh := range shutdownChToClose {
+		close(shutdownCh)
+	}
+}
+
+func (s *SequentialScheduler[T]) pollTaskQueue(workerShutdownCh <-chan struct{}) {
+	defer s.shutdownWG.Done()
+
+	for {
+		select {
+		case <-s.shutdownChan:
+			return
+		case <-workerShutdownCh:
+			return
+		case queue := <-s.queueChan:
+			s.processTaskQueue(queue, workerShutdownCh)
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) processTaskQueue(
+	queue SequentialTaskQueue[T],
+	workerShutdownCh <-chan struct{},
+) {
+	for {
+		select {
+		case <-s.shutdownChan:
+			// Put queue back to the queue channel
+			s.queueChan <- queue
+			return
+		case <-workerShutdownCh:
+			// Put queue back to the queue channel
+			s.queueChan <- queue
```

Comment:
compared to queue shutdown, worker shutdown should give the queue back to scheduler